### ISSUE TITLE
FSDP do not support `ignored_parameters` when `auto_wrap_policy` is specified

### ISSUE THREAD

**Issue by KimmiShi** [pytorch/pytorch#109100]
### 🐛 Describe the bug

I have a model which contains some params need to be ignored ( else flat_param will raise an error), the construction code is like:

```
        not_trainable = []
        for name, p in model.named_parameters():
            if p.requires_grad == False:
                not_trainable.append(p)
        sharding_strategy=torch.distributed.fsdp.ShardingStrategy._HYBRID_SHARD_ZERO2
        model = FSDP(model, sharding_strategy=sharding_strategy,
                     ignored_parameters = not_trainable,
                     auto_wrap_policy=size_based_auto_wrap_policy)

```
I came across an error: the params in `not_trainable ` would cause error when executing `_init_flat_param`.

If `auto_wrap_policy` is not passed, the code works.

### Versions

PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 10.2.0

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin

**Comment by awgu** [comment]
`ignored_parameters` got renamed to `ignored_states` in the latest `main`. There are also some known bugs around `ignored_states` if you ignore only some parameters but not all of a module.

To unblock in the short term, you can either use `use_orig_params=True` and use higher gradient memory, or you can rewrite your `nn.Module` to group the non-trainable parameters into separate modules.

**Comment by KimmiShi** [comment]
Hi @awgu, 

I am a little bit confused by the behavior of FSDP:

when I use `SHARD_GRAD_OP`, the `model.parameters()` returns sharded params, but `model.state_dict` returns full params. If the params are sharded, how could `model.state_dict` be full?

I am wondering if `model.state_dict` returns valid state_dict, is it the same value with `model.paramters`? 

Thanks a lot!

**Comment by awgu** [comment]
@KimmiShi If `model.state_dict()` returns full parameters, then that means that you are saving a full model state dict, which is the default right now.
https://github.com/pytorch/pytorch/blob/fc47ba2794564592a243b8b3484b5cfee1e72fda/torch/distributed/fsdp/_init_utils.py#L511

You can use `set_state_dict_type()` to configure this:
https://github.com/pytorch/pytorch/blob/fc47ba2794564592a243b8b3484b5cfee1e72fda/torch/distributed/fsdp/fully_sharded_data_parallel.py#L803-L811

`model.state_dict()` does not have the same values as `model.parameters()`.
